packages/financial-templates-lib/src/price-feed/utils.js

```diff
@@ -168,6 +168,97 @@ exports.PriceHistory = (getPrice, prices = {}) => {
   };
 };
 
+/**
+ * @notice Constructs the BlockFinder, which is a simple class-like object to find blocks by timestamp, but keeps a
+ *      cache to optimize the search.
+ * @param {Function} requestBlock async function, like web3.eth.getBlock that returns a block for a particular block
+ *      number or returns the latest block if no argument is provided. Blocks returned must have a `number` and
+ */
+exports.BlockFinder = (requestBlock, blocks = []) => {
+  // Grabs the most recent block and caches it.
+  async function getLatestBlock() {
+    const block = await requestBlock();
+    const index = lodash.sortedIndexBy(blocks, block, "number");
+    if (blocks[index]?.number !== block.number) blocks.splice(index, 0, block);
+    return blocks[index];
+  }
+
+  // Grabs the block for a particular number and caches it.
+  async function getBlock(number) {
+    const index = lodash.sortedIndexBy(blocks, { number }, "number");
+    if (blocks[index]?.number === number) return blocks[index]; // Return early if block already exists.
```

Comment:
seeing you using the `?.` syntax, if this is not the type you expect you might have some issues. do you want to sanitize the inputs before, as with the other methods?

Reply:
This is to deal with the case where index doesn't exist, which means blocks[index] === undefined. This happens (I believe) if the block number should be slotted at the end of the blocks array.

Reply:
Two other comments:
- blocks isn't an input, so I don't think sanitizing them would give us guarantees about the types in that array.
- Regarding sanitizing inputs, generally, we _could_ ensure that number is defined, but since it is only called internally (function isn't exported), I'm somewhat conflicted on whether we should check types when we can trivially see all the callsites nearby, but I don't have strong opinions on that!

Reply:
sounds good to me!